libibverbs/device.c

```diff
@@ -222,6 +222,44 @@ static bool has_ioctl_write(struct ibv_context *ctx)
 	return true;
 }
 
+static void verbs_set_log_level(struct verbs_context *context_ex)
+{
+	char *env;
+
+	env = getenv("VERBS_LOG_LEVEL");
+	if (env)
+		context_ex->log_level = strtol(env, NULL, 0);
+}
+
+static void verbs_set_log_file(struct verbs_context *context_ex)
+{
+	char *env;
+
+	if (context_ex->log_level == VERBS_LOG_LEVEL_NONE)
+		return;
+
+	env = getenv("VERBS_LOG_FILE");
```

Comment:
The mlx5 implementation is slightly different and IMHO better.

The mlx5 prints are visible only if library is compiled with MLX5_DEBUG flag, without such special compilation flag, nothing is printed to the screen. This allows us to use logging both in data and control paths. Yes, "if ((level) <= (ctx)->log_level)" costs us enough to do not put it in hot paths.

The FD is needed where you anyway want to print something, but the library is not compiled with MLX5_DEBUG.

I think that you can rely CMAKE_BUILD_TYPE to provide same behavior.

Reply:
mlx5 wraps some of the prints with the MLX5_DEBUG define, you can still do that for prints in data path (or use the CMAKE_BUILD_TYPE).
I think it's better for control path prints to not be dependent on such define, this way they will always be available.

Reply:
You need two level macros/defines anyway. One for control path, another for data path. Let's add data path macros too.

Reply:
What do you mean?
I meant that in data path you can wrap the verbs_* calls with #ifdefs, do you want a version of verbs_* that does that internally?

Reply:
Something like that, but I still didn't convince myself that debug prints for library compiled as "Release" is right thing to do.

Reply:
Taking mlx5 as an example, the mlx5_err() prints do not depend on how the library is compiled, the mlx5_dbg() does.
I can do something similar where verbs_debug() only prints in debug mode and other stuff are always available.

Reply:
It is not always available, but only if you provided DEBUG_FILE

Reply:
You're right, it always checks the if statement but doesn't print if there's no DEBUG_FILE.
So do you have any preferences how to do it here?

Reply:
I would expect two modes:
1. library compiled in debug mode - print whatever you want without any need in DEBUG_FILE
2. library compiled in release mode - request from the user to supply extra variable VERBS_DEBUG_FILE. it will be super nice if we will be civil and provide a way to use syslog in addition/instead of debug file. Or at least will leave a room to implement it later.

Reply:
I thought the motivation was to not run the additional if statement when compiled in release mode, I'm not sure we can do that with what you described?

Reply:
> I thought the motivation was to not run the additional if statement when compiled in release mode, I'm not sure we can do that with what you described?

Yes, this is the motivation, this "if" will be for verbs_err() macro only. All other will be compiled out. Maybe even special macro: verbs_err_datapath()